crypto/context.c

```diff
@@ -92,19 +174,88 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
 
     if (ctx == NULL) {
         if (!RUN_ONCE(&default_context_init, do_default_context_init))
-            return 0;
-        ctx = &default_context;
+            return NULL;
+        ctx = default_context;
     }
+    if (ctx == NULL)
+        return NULL;
 
     CRYPTO_THREAD_read_lock(ctx->lock);
 
     /* The alloc call ensures there's a value there */
     if (CRYPTO_alloc_ex_data(CRYPTO_EX_INDEX_OPENSSL_CTX, NULL,
-                             &ctx->data, index))
-        data = CRYPTO_get_ex_data(&ctx->data, index);
+                             &ctx->data, ctx->dyn_indexes[index]))
+        data = CRYPTO_get_ex_data(&ctx->data, ctx->dyn_indexes[index]);
 
     CRYPTO_THREAD_unlock(ctx->lock);
 
     return data;
 }
 
+OSSL_EX_DATA_GLOBAL *openssl_ctx_get_ex_data_global(OPENSSL_CTX *ctx)
+{
+#ifndef FIPS_MODE
+    if (ctx == NULL) {
+        if (!RUN_ONCE(&default_context_init, do_default_context_init))
+            return NULL;
+        ctx = default_context;
+    }
+#endif
```

Comment:
I'm not sure that we need to make this explicitly non-FIPS

Reply:
Its dead code in FIPS since we will never have a default context on the provider side there so I'd prefer to compile out the RUN_ONCE path altogether.

Reply:
If it's dead code it's kinda harmless and I think you should consider "minimize ifdefs" as a goal as well, even if you come to the same conclusion.

Reply:
Well let me put it another way. It *should* be dead code in the FIPS provider. In the FIPS provider `ctx` should never be NULL on entry to this function. If for some reason a bug means that we enter this function and it *is* NULL then with this ifdef there we will fail immediately. Without it there we will continue on, it will automatically allocate a ctx completely disconnected from the ctx it is supposed to be using and weird stuff will happen - probably failing in some completely unrelated location - making debugging very hard.

Reply:
I kinda suspected that's what you wanted.  Please add commentary on the subject if you haven't already (I haven't looked for that...)

Reply:
Comment added